source/common/network/win32_redirect_records_option_impl.cc

```diff
@@ -0,0 +1,77 @@
+#include "common/network/win32_redirect_records_option_impl.h"
+
+#include "envoy/common/exception.h"
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/common/scalar_to_byte_vector.h"
+#include "common/common/utility.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+namespace {
+// `IOCTL` controls unlike socket options do not have level parameter. So we arbitrarily define one
+// in Envoy.
+#define ENVOY_IOCTL_LEVEL 1
+
+// We create an Envoy definition for `SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS` to make the
+// unit tests work on both platforms.
+#ifdef SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS
+#define ENVOY_SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS
+#else
+#define ENVOY_SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS -1
+#endif
+
+Network::SocketOptionName getWfpRedirectionOption() {
+  return ENVOY_MAKE_SOCKET_OPTION_NAME(ENVOY_IOCTL_LEVEL,
+                                       ENVOY_SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS);
+}
+} // namespace
+
+const Network::SocketOptionName& Win32RedirectRecordsOptionImpl::optionName() {
+  CONSTRUCT_ON_FIRST_USE(Network::SocketOptionName, getWfpRedirectionOption());
+}
+
+// Socket::Option
+bool Win32RedirectRecordsOptionImpl::setOption(
+    Socket& socket, envoy::config::core::v3::SocketOption::SocketState state) const {
+  if (in_state_ == state) {
+    unsigned long size = 0;
+    const Api::SysCallIntResult result =
+        socket.win32Ioctl(ENVOY_SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS,
+                          const_cast<void*>(reinterpret_cast<const void*>(redirect_records_.buf_)),
+                          redirect_records_.buf_size_, nullptr, 0, &size);
+    if (result.rc_ != 0) {
+      ENVOY_LOG(warn, "Setting WFP records on socket failed: {}", errorDetails(result.errno_));
```

Comment:
Consider always logging even for success case (with result.rc value) if caller doesn't.

Reply:
The caller logs if there is a failure in applying socket options and the rest of the socket options implementing this interface are only logging failures. From what I have seen the common way is to only log failures

Reply:
I think it's fine to only log failure. If non-failure logging were to be added, it should be at `trace` level.